File: app/lib/utils/errorHandler.ts
/**
 * Centralized error handler for the React/Next.js web app.
 * Maps raw Supabase / network exceptions to user-friendly messages.
 * Import and call `friendlyError(error)` wherever you catch exceptions.
 */

export function friendlyError(error: unknown): string {
    if (!error) return 'Something went wrong. Please try again.';

    const msg =
        (error instanceof Error ? error.message : String(error)).toLowerCase();

    // ── Network ────────────────────────────────────────────────
    if (
        msg.includes('failed to fetch') ||
        msg.includes('networkerror') ||
        msg.includes('network') ||
        msg.includes('connection')
    ) {
        return 'No internet connection. Please check your network.';
    }

    if (msg.includes('timeout') || msg.includes('timed out')) {
        return 'The request timed out. Please try again.';
    }

    // ── Supabase Auth ──────────────────────────────────────────
    if (msg.includes('invalid login credentials') || msg.includes('invalid email or password')) {
        return 'Incorrect email or password. Please try again.';
    }
    if (msg.includes('user already registered') || msg.includes('already been registered')) {
        return 'An account with this email already exists.';
    }
    if (msg.includes('email not confirmed')) {
        return 'Please verify your email before logging in.';
    }
    if (msg.includes('password') && msg.includes('weak')) {
        return 'Password must be at least 6 characters.';
    }
    if (msg.includes('signup_disabled')) {
        return 'New registrations are currently disabled.';
    }
    if (msg.includes('jwt') || msg.includes('session_not_found')) {
        return 'Your session has expired. Please log in again.';
    }

    // ── Supabase DB / Storage ──────────────────────────────────
    if (msg.includes('duplicate') || msg.includes('unique')) {
        return 'This already exists. No duplicate entries allowed.';
    }
    if (msg.includes('foreign key') || msg.includes('not-null')) {
        return 'Some required information is missing.';
    }
    if (msg.includes('permission denied') || msg.includes('rls')) {
        return 'You do not have permission to do that.';
    }
    if (msg.includes('storage') || msg.includes('bucket')) {
        return 'File upload failed. Please try a different file.';
    }
    if (msg.includes('too large') || msg.includes('payload too large')) {
        return 'The file is too large. Please choose a smaller one.';
    }

    // ── HTTP server errors ─────────────────────────────────────
    if (msg.includes('500') || msg.includes('internal server')) {
        return 'Server error. Please try again later.';
    }
    if (msg.includes('503') || msg.includes('service unavailable')) {
        return 'Service temporarily unavailable. Try again shortly.';
    }
    if (msg.includes('401') || msg.includes('unauthorized')) {
        return 'You need to be logged in to do that.';
    }
    if (msg.includes('403') || msg.includes('forbidden')) {
        return 'You do not have permission to do that.';
    }
    if (msg.includes('404') || msg.includes('not found')) {
        return 'The requested resource could not be found.';
    }

    // ── Fallback ───────────────────────────────────────────────
    return 'Something went wrong. Please try again.';
}

File: mentron_flutter/lib/core/utils/error_handler.dart
/// Centralized error handler — maps raw exceptions to user-friendly messages.
/// Never shows technical stack traces or internal error codes to the user.
class ErrorHandler {
  /// Returns a clean, friendly message from any caught exception.
  static String friendly(Object? error) {
    if (error == null) return 'Something went wrong. Please try again.';

    final msg = error.toString().toLowerCase();

    // ── Network / Connection ──────────────────────────────────
    if (msg.contains('socketexception') ||
        msg.contains('failed host lookup') ||
        msg.contains('network') ||
        msg.contains('connection') ||
        msg.contains('internet')) {
      return 'No internet connection. Please check your network and try again.';
    }

    // ── Timeout ───────────────────────────────────────────────
    if (msg.contains('timeout') || msg.contains('timed out')) {
      return 'The request timed out. Please try again.';
    }

    // ── Supabase Auth ─────────────────────────────────────────
    if (msg.contains('invalid login credentials') ||
        msg.contains('invalid email or password')) {
      return 'Incorrect email or password. Please try again.';
    }
    if (msg.contains('user already registered') ||
        msg.contains('already been registered')) {
      return 'An account with this email already exists.';
    }
    if (msg.contains('email not confirmed')) {
      return 'Please verify your email before logging in.';
    }
    if (msg.contains('weak password') || msg.contains('password')) {
      return 'Password must be at least 6 characters long.';
    }
    if (msg.contains('signup_disabled')) {
      return 'New registrations are currently disabled.';
    }
    if (msg.contains('jwt') || msg.contains('session')) {
      return 'Your session has expired. Please log in again.';
    }

    // ── Supabase Database / Storage ───────────────────────────
    if (msg.contains('violates') || msg.contains('duplicate')) {
      return 'This record already exists.';
    }
    if (msg.contains('foreign key') || msg.contains('not-null')) {
      return 'Some required information is missing.';
    }
    if (msg.contains('permission denied') || msg.contains('rls')) {
      return 'You do not have permission to do that.';
    }
    if (msg.contains('storage') || msg.contains('bucket')) {
      return 'File upload failed. Please try a different file.';
    }
    if (msg.contains('too large') || msg.contains('payload')) {
      return 'The file is too large. Please choose a smaller file.';
    }

    // ── Server errors ─────────────────────────────────────────
    if (msg.contains('500') || msg.contains('server error')) {
      return 'Server error. Please try again later.';
    }
    if (msg.contains('503') || msg.contains('unavailable')) {
      return 'Service temporarily unavailable. Try again shortly.';
    }

    // ── Fallback ──────────────────────────────────────────────
    return 'Something went wrong. Please try again.';
  }
}
